File: Cryptography/HW5/primrootcheck.java
import java.math.BigInteger;
import java.util.ArrayList;
import java.util.Scanner;
class PrimRootCheck {
    public static void main(String[] args) {
        //Variable setup and setup code for algorithms
        int n;
        BigInteger p = new BigInteger("0");
        BigInteger g = new BigInteger("0");
        Scanner s = new Scanner(System.in);

        n = s.nextInt();

        for(int i = 0; i < n; i++){
            p = s.nextBigInteger();
            g = s.nextBigInteger();

            if(isPrimitiveRoot(p, g) == true){
                System.out.println(1);
            }
            else {
                System.out.println(0);
            }
        }

        s.close();
    }

    public static boolean isPrimitiveRoot(BigInteger p, BigInteger g){
        //First we need to determine the divisors of p
        ArrayList<BigInteger> factors = new ArrayList<>();

        BigInteger check = p.subtract(BigInteger.ONE);
        //In my proof for Number 7, I showed that we only need to look at factors that are prime numbers
        //This is because that is where a non-primitive roots' cycles will first appear.
        //However, for the sake of this problem, I will check all factors since i am kinda lazy
        BigInteger i = BigInteger.TWO;
        while(true){
            if(i.multiply(i).compareTo(check) == 1){
                break;
            }

            if(check.mod(i).equals(BigInteger.ZERO)){
                //Add both the number and the value that multiplies said number to p - 1
                factors.add(i);
                factors.add(check.divide(i));
            }

            i = i.add(BigInteger.ONE);
        }

        long numFactors = factors.size();

        //If we have a remainder of 1, there is a cycle therefore g is not a primitive root of p
        for(int j = 0; j < numFactors; j++){
            if(modExp(g, check.divide(factors.get(j)), p).equals(BigInteger.ONE)){
                return false;
            }
        }
        return true;
    }

    //Make a fast exponentiation algorithm
    public static BigInteger modExp(BigInteger num, BigInteger power, BigInteger mod){
        if(power.equals(BigInteger.ZERO)) return BigInteger.ONE;
        if(power.equals(BigInteger.ONE)) return num.mod(mod);
        if(power.mod(BigInteger.TWO).equals(BigInteger.ZERO)){
            BigInteger q = modExp(num, power.divide(BigInteger.TWO), mod);
            return (q.multiply(q)).mod(mod);
        }
        return (num.multiply(modExp(num, power.subtract(BigInteger.ONE), mod)).mod(mod));
    }
}

File: Cryptography/HW5/primRootCheck.cpp
#include <vector>
#include <iostream>

long long modMult(long long x, long long y, long long mod){
    long long q = (mod / x);
    return x * (y % q) - (mod % x) * (y / q) < 0 ? (x * (y % q) - (mod % x) * (y / q) + mod) % mod : (x * (y % q) - (mod % x) * (y / q)) % mod;
}

long long modExp(long long num, long long power, long long mod){

    if(power == 0) return 1;
    if(power == 1) return num % mod;
    if(power % 2 == 0){
        long long q = modExp(num, power/2, mod);
        long long val = modMult(q, q, mod) % mod;

        std::cout << val << std::endl;
        return (val);
    }
    return ((num % mod) * modExp(num, power - 1, mod)) % mod;
}

bool isPrimitiveRoot(long long p, long long g, std::vector<long long> &factors){


    for(long long i = 2; i * i < p - 1; i++){
        if((p - 1) % i == 0){
            factors.push_back(i);
            factors.push_back((p - 1) / i);
        }
    }
    long long numFactors = factors.size();

    for(long long i = 0; i < numFactors; i++){
        if(modExp(g, (p - 1) / factors[i], p) == 1){
            return false;
        }
    }
    return true;
}

int main(){
    int n;
    long long p, g;

    std::vector<long long> factors;

    std::cin >> n;

    for(int i = 0; i < n; i++){
        std::cin >> p >> g;

        if(isPrimitiveRoot(p, g, factors) == true){
            std::cout << 1 << std::endl;
        }
        else {
            std::cout << 0 << std::endl;
        }

        factors.clear();
    }
}